src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.AddSub.cs

```diff
@@ -9,58 +9,131 @@ namespace System.Numerics
 {
     internal static partial class BigIntegerCalculator
     {
+
+#if DEBUG
+        // Mutable for unit testing.
+        private static
+#else
+        private const
+#endif
+        int CopyToThreshold = 8;
+
+        private static void CopyTail(ReadOnlySpan<uint> source, Span<uint> dest, int start)
+        {
+            source.Slice(start).CopyTo(dest.Slice(start));
+        }
+
         public static void Add(ReadOnlySpan<uint> left, uint right, Span<uint> bits)
         {
             Debug.Assert(left.Length >= 1);
             Debug.Assert(bits.Length == left.Length + 1);
 
+            // Switching to managed references helps eliminating
+            // index bounds check...
```

Comment:
The previous version of the code was eliminating index bound checks for `left[i]`, as it was always used within a loop that was recognized by the JIT:

```cs
for (int i = 0; i < left.Length; i++)
   long digit = left[i] + carry;
```

This was not true for `bits`, because JIT did not know that `bits.Length == left.Length + 1`.

We could avoid using managed references for `left` and just keep using the old loop here and access `bits` via `Unsafe.Add(ref managedRefToBits)`.

```suggestion
            // Switching to managed references helps eliminating
            // index bounds check for both buffers.
```